include/picongpu/plugins/openPMD/WriteSpecies.hpp

```diff
@@ -256,25 +295,91 @@ namespace picongpu
+                       ::openPMD::UnitDimension::I,
+                       ::openPMD::UnitDimension::theta,
+                       ::openPMD::UnitDimension::N,
+                       ::openPMD::UnitDimension::J};
+                std::map<::openPMD::UnitDimension, double> zeroUnitMap;
+                for(unsigned i = 0; i < 7; ++i)
+                {
+                    zeroUnitMap[openPMDUnitDimensions[i]] = zeroUnitDimension[i];
+                }
+                // end of do-not-touch code
+
+                // now we have a map in a writeable format with all zeroes
+                // for each record copy it and modify the copy, e.g.
+
+                // const records stuff
+                ::openPMD::Datatype dataType = ::openPMD::Datatype::DOUBLE;
+                ::openPMD::Extent extent = {0};
+                ::openPMD::Dataset dataSet = ::openPMD::Dataset(dataType, extent);
+
+                // mass
+                ::picongpu::detail::GetMassOrZero<FrameType> const getMassOrZero;
+                if(getMassOrZero.hasMassRatio)
+                {
+                    const float_64 mass(getMassOrZero());
+                    auto& massRecord = record["mass"];
+                    auto& massComponent = massRecord[::openPMD::RecordComponent::SCALAR];
+                    massComponent.resetDataset(dataSet);
+                    massComponent.makeConstant(mass);
+
+                    auto massUnitMap = zeroUnitMap;
+                    massUnitMap[::openPMD::UnitDimension::M] = 1.0;
+                    massRecord.setUnitDimension(massUnitMap);
+                    massComponent.setUnitSI(::picongpu::UNIT_MASS);
+                    massRecord.setAttribute("macroWeighted", int32_t(false));
+                    massRecord.setAttribute("weightingPower", float_64(1.0));
+                    massRecord.setAttribute("timeOffset", float_64(0.0));
+                }
+
+
+                // charge
+                using hasBoundElectrons = typename pmacc::traits::HasIdentifier<FrameType, boundElectrons>::type;
+                ::picongpu::detail::GetChargeOrZero<FrameType> const getChargeOrZero;
+                if(!hasBoundElectrons::value && getChargeOrZero.hasChargeRatio)
+                {
+                    const float_64 charge(getChargeOrZero());
+                    auto& chargeRecord = record["charge"];
+                    auto& chargeComponent = chargeRecord[::openPMD::RecordComponent::SCALAR];
+                    chargeComponent.resetDataset(dataSet);
+                    chargeComponent.makeConstant(charge);
+
+                    auto chargeUnitMap = zeroUnitMap;
+                    chargeUnitMap[::openPMD::UnitDimension::I] = 1.0;
+                    chargeUnitMap[::openPMD::UnitDimension::T] = 1.0;
+                    chargeRecord.setUnitDimension(chargeUnitMap);
+                    chargeComponent.setUnitSI(::picongpu::UNIT_CHARGE);
+                    chargeRecord.setAttribute("macroWeighted", int32_t(false));
```

Comment:
note: only momentum is `macroWeighted = 1` in PIConGPU. potentially you need to add that case here.

Reply:
This part does not handle momentums. It basically handles whatever attributes were missed in the original implementation which was covering momentum

Reply:
I think @ax3l is right this PR is handling only charge and mass but not all other attributes.